File: MergeSort.cpp
#include<iostream>
#include<vector>

using namespace std;

void display(vector<int> arr)
{
    for(int i=0;i<arr.size();i++)
        cout<<arr[i]<<"\t";
    cout<<endl;
}

vector<int> merge(vector<int> &arr1, vector<int>&arr2)
{
    vector<int> temp;
    int a1=0,a2=0;
    while(a1<arr1.size() && a2<arr2.size())
    {
        if(arr1[a1]<arr2[a2])
        {
            temp.push_back(arr1[a1++]);
        }
        else
        {
            temp.push_back(arr2[a2++]);
        }
    }
    while(a1<arr1.size())
        temp.push_back(arr1[a1++]); 
    while(a2<arr2.size())
        temp.push_back(arr2[a2++]);
    return temp;
}

vector<int> mergeSort(vector<int> arr, int low, int high)
{
    // cout<<"Mergesort rec low : "<<low<<" high : "<<high<<endl;
    vector<int> a1;
    if(low<high)
    {
        int mid = (low+high)/2;
        // int k;
        // cin>>k;
        // cout<<"Calling mergesort : "<<low<<" to "<<mid<<endl;
        vector<int> a2 = mergeSort(arr,low,mid);
        // display(a2);
        // cout<<"Calling mergesort : "<<mid+1<<" to "<<high<<endl;
        vector<int> a3 = mergeSort(arr,mid+1,high);
        // display(a3);
        // cout<<"Calling merge : "<<low<<" to "<<high<<endl;
        a1 = merge(a2,a3);
        // display(a1);
    }
    if(low==high)
    {
        a1.push_back(arr[low]);
        return a1;
    }
    return a1;
}


int main()
{
    vector<int> arr1 = {1, 2, 3, 5, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 22, 23, 24, 25, 26, 27, 28, 29, 31, 32, 33, 34, 35, 37, 39, 40, 41, 42, 43, 45, 46, 50, 51, 52, 54, 55, 57, 58, 59, 60};
    vector<int> arr3 = mergeSort(arr1,0,arr1.size()-1);
    display(arr3);

    return 0;
}

File: MergeSort.py

def merge(lst,l,m,r):
    n1 = m-l+1
    n2 = r-m
    ll = []
    rl = []

    for i in range(n1):
        ll.append(lst[l+i])
    for i in range(n2):
        rl.append(lst[m+i+1])
    i=0
    j=0
    k=l
    while k<=r and i<n1 and j<n2:
        if ll[i]<=rl[j]:
            lst[k]=ll[i]
            i+=1
        else:
            lst[k]=rl[j]
            j+=1
        k+=1

    while k<=r:
        if i<n1:
            lst[k]=ll[i]
            k+=1
            i+=1
        if j<n2:
            lst[k]=rl[j]
            k+=1
            j+=1


def mergeSort(lst,l,r):
    m = int((l+r)/2)
    if (r-l)>0:
        mergeSort(lst,l,m)
        mergeSort(lst,m+1,r)
        merge(lst,l,m,r)


lst = [710,827,383,986,350,259,135,595,890,44,776,489,408,437,218,738,334,394,954,418,925,922,778,777,716,563,252,110,556,538,97,707,542,792,614,613,885,257,961,266,652,604,89,118,390,609,736,449,268,639,170,7,142,761,398,382,267,642,943,530,169,173,920,292,602,945,393,234,969,822]
mergeSort(lst,0,len(lst)-1)
print(lst)






